File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/client/AccountAliasResolution.java
package org.hyperledger.iroha.android.client;

import java.math.BigInteger;
import org.hyperledger.iroha.android.address.AccountIdLiteral;
import org.hyperledger.iroha.android.alias.AccountAliasName;

/** Parsed payload for account alias resolution (`/v1/aliases/resolve`). */
public final class AccountAliasResolution {
  private final String alias;
  private final String accountId;
  private final BigInteger index;
  private final String source;

  public AccountAliasResolution(
      final String alias, final String accountId, final BigInteger index, final String source) {
    final String canonicalAlias = AccountAliasName.parse(alias).canonicalText();
    if (!canonicalAlias.equals(alias)) {
      throw new IllegalArgumentException("alias must use its canonical representation");
    }
    this.alias = canonicalAlias;
    this.accountId = AccountIdLiteral.requireCanonicalI105Address(accountId, "accountId");
    this.index = index == null ? null : AccountAliasUInt64.require(index, "index");
    this.source = source == null ? null : AccountAliasesByAccount.requireExactText(source, "source");
  }

  public String alias() {
    return alias;
  }

  public String accountId() {
    return accountId;
  }

  public BigInteger index() {
    return index;
  }

  public String source() {
    return source;
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/client/AccountAliasResolution.kt
package org.hyperledger.iroha.sdk.client

import java.math.BigInteger
import org.hyperledger.iroha.sdk.address.requireCanonicalI105Address
import org.hyperledger.iroha.sdk.alias.AccountAliasName

/** Parsed payload for account alias resolution (`/v1/aliases/resolve`). */
class AccountAliasResolution(
    alias: String,
    accountId: String,
    index: BigInteger?,
    source: String?,
) {
    /** Exact canonical account alias returned by Torii. */
    @JvmField
    val alias: String = AccountAliasName.parse(alias).canonicalText().also {
        require(it == alias) { "alias must use its canonical representation" }
    }

    /** Exact canonical domainless account identifier returned by Torii. */
    @JvmField
    val accountId: String = requireCanonicalI105Address(accountId, "accountId")

    /** Optional numeric alias index. */
    @JvmField
    val index: BigInteger? = index?.let { requireAliasU64(it, "index") }

    /** Resolution source, when supplied by Torii. */
    @JvmField
    val source: String? = source?.also { requireExactText(it, "source") }
}
